lib/pgBackRest/Storage/Posix/Driver.pm

```diff
@@ -0,0 +1,864 @@
+    my
+    (
+        $strOperation,
+        $strSourcePath,
+        $strDestinationPath,
+        $bHard,
+        $bPathCreate,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->linkCreate', \@_,
+            {name => 'strSourcePath', trace => true},
+            {name => 'strDestinationPath', trace => true},
+            {name => 'bHard', optional=> true, default => false, trace => true},
+            {name => 'bPathCreate', optional=> true, default => true, trace => true},
+        );
+
+    if (!($bHard ? link($strSourcePath, $strDestinationPath) : symlink($strSourcePath, $strDestinationPath)))
+    {
+        my $strMessage = "unable to create link '${strDestinationPath}'";
+
+        # If parent path or source is missing
+        if ($OS_ERROR{ENOENT})
+        {
+            # Check if source is missing
+            if (!$self->exists($strSourcePath))
+            {
+                confess &log(ERROR, "${strMessage} because source '${strSourcePath}' does not exist", ERROR_FILE_MISSING);
+            }
+
+            if (!$bPathCreate)
+            {
+                confess &log(ERROR, "${strMessage} because parent does not exist", ERROR_PATH_MISSING);
+            }
+
+            # Create parent path
+            $self->pathCreate(dirname($strDestinationPath), {bIgnoreExists => true, bCreateParent => true});
+
+            # Create link
+            $self->linkCreate($strSourcePath, $strDestinationPath, {bHard => $bHard});
+        }
+        # Else if link already exists
+        elsif ($OS_ERROR{EEXIST})
+        {
+            confess &log(ERROR, "${strMessage} because it already exists", ERROR_PATH_EXISTS);
+        }
+        else
+        {
+            logErrorResult(ERROR_PATH_CREATE, ${strMessage}, $OS_ERROR);
+        }
+    }
+
+    # Return from function and log return values if any
+    return logDebugReturn($strOperation);
+}
+
+####################################################################################################################################
+# move - move a file.
+####################################################################################################################################
+sub move
+{
+    my $self = shift;
+
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $strSourceFile,
+        $strDestinationFile,
+        $bPathCreate,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->move', \@_,
+            {name => 'strSourceFile', trace => true},
+            {name => 'strDestinationFile', trace => true},
+            {name => 'bPathCreate', default => false, trace => true},
+        );
+
+    # Get source and destination paths
+    my $strSourcePath = dirname($strSourceFile);
+    my $strDestinationPath = dirname($strDestinationFile);
+
+    # Move the file
+    if (!rename($strSourceFile, $strDestinationFile))
+    {
```

Comment:
Move are used to ensure atomicity on Posix so we need to use the low level move.